modules/renter/siafile/siafile.go

```diff
@@ -0,0 +1,181 @@
+package siafile
+
+import (
+	"bytes"
+	"encoding/base32"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/fastrand"
+
+	"github.com/NebulousLabs/Sia/crypto"
+)
+
+type (
+	// SiaFile is the disk format for files uploaded to the Sia network.  It
+	// contains all the necessary information to recover a file from its hosts and
+	// allows for easy constant-time updates of the file without having to read or
+	// write the whole file.
+	SiaFile struct {
+		// metadata is the mostly static metadata of a SiaFile. The reserved
+		// size of the metadata on disk should always be a multiple of 4kib.
+		// The metadata is also the only part of the file that is JSON encoded
+		// and can therefore be easily extended.
+		metadata Metadata
+
+		// pubKeyTable stores the public keys of the hosts this file's pieces are uploaded to.
+		// Since multiple pieces from different chunks might be uploaded to the same host, this
+		// allows us to deduplicate the rather large public keys.
+		pubKeyTable []types.SiaPublicKey
+
+		// chunks are the chunks the file was split into.
+		chunks []Chunk
+
+		// utility fields. These are not persisted.
+		deleted     bool
+		erasureCode modules.ErasureCoder
+		mu          sync.RWMutex
+		uid         string
+	}
+
+	// Chunk represents a single chunk of a file on disk
+	Chunk struct {
+		// erasure code settings.
+		//
+		// erasureCodeType specifies the algorithm used for erasure coding
+		// chunks. Available types are:
+		//   0 - Invalid / Missing Code
+		//   1 - Reed Solomon Code
+		//
+		// erasureCodeParams specifies possible parameters for a certain
+		// erasureCodeType. Currently params will be parsed as follows:
+		//   Reed Solomon Code - 4 bytes dataPieces / 4 bytes parityPieces
+		//
+		erasureCodeType   [4]byte
+		erasureCodeParams [8]byte
+
+		// extensionInfo is some reserved space for each chunk that allows us
+		// to indicate if a chunk is special.
+		extensionInfo [16]byte
+
+		// pieces are the pieces of the file the chunk consists of.
+		pieces [][]Piece
+	}
+
+	// Piece represents a single piece of a chunk on disk
+	Piece struct {
+		KeyNonce   [4]byte            // nonce used for encrypting the piece
+		HostPubKey types.SiaPublicKey // public key of the host
+		MerkleRoot crypto.Hash        // merkle root of the piece
+	}
```

Comment:
so, this is something we aren't going to be able to do soon, but I do think that we are going to at some point need a mapping in the contractor of all of the pieces that are in each contract. it'll probably need to be on disk, but it's the only way I can think of to be sure that we actually have the data still in the contract, especially once we build in the rpc to allow hosts to tell us that they've lost some pieces (in the event of a corrupted disk, for example) but not all of them in the contract.

Once we have that mapping, we genuinely won't need 32 bytes of the merkle root per piece, we'll be able to safely trim it down to about 10 bytes and still have high confidence of correctness, because every piece in our contracts was of our own choosing, and if there's ever a conflict / collision we can just change the key nonce for the piece and try again.

Reply:
On the note of key nonces, a disadvantage of this strategy is that you can't modify only part of a chunk, if you are going to modify a file on Sia you have to modify the whole chunk to stay secure. I'm not sure if there are encryption strategies that allow you to perform safe in-place modification without needing a ton of nonce overhead though.

Reply:
Doesn't XTS let you do that without a nonce? I think it is used by Microsoft in Bitlocker.